### PR TITLE
3444 - 0

### PR DESCRIPTION
**Pull request by svintbel**
## Номер
3444

## Номер задания
0

## Ссылка на видео с демо
https://youtu.be/O09-Hci9H_k

## Комментарии
No comments

### CODE REVIEW

File: 3444/0/about_scoring_project.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  if dice.length > 2
+    (1..6).each do |number|
+      next unless dice.count(number) > 2
+
+      total_score += triple_value[number]
+      3.times { dice.delete_at(dice.index(number)) }
+      break
+    end
+  end
+  dice.each do |number|
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: use `sum`

File: 3444/0/about_scoring_project.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+# rubocop:disable Metrics/MethodLength
+
+# :reek:UncommunicativeVariableName and :reek:TooManyStatements and :reek:UtilityFunction
+def score(dice)
+  total_score = 0
+  triple_value = [nil, 1000, 200, 300, 400, 500, 600]
+  single_value = [nil, 100, 0, 0, 0, 50, 0]
+  if dice.length > 2
+    (1..6).each do |number|
+      next unless dice.count(number) > 2
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: use if

File: 3444/0/about_scoring_project.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+      total_score += triple_value[number]
+      3.times { dice.delete_at(dice.index(number)) }
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: try to avoid arguments mutation

File: 3444/0/about_scoring_project.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      break
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: break? O_o. In 99.99% of cases you don't really need it. Look at `detect` method for example

File: 3444/0/about_scoring_project.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+# score([1,5,1,2,4]) => 250 points
+#
+# More scoring examples are given in the tests below:
+#
+# Your goal is to write the score method.
+# rubocop:disable Metrics/MethodLength
+
+# :reek:UncommunicativeVariableName and :reek:TooManyStatements and :reek:UtilityFunction
+def score(dice)
+  total_score = 0
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: well it's really very complex solution. Looks like magic. Try to simplify ur code. Maybe even try different approach.
You can try to use [`tally`](https://medium.com/@baweaver/ruby-2-7-enumerable-tally-a706a5fb11ea) if you are on 2.7

File: 3444/0/triangle.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#
+
+# :reek:UtilityFunction and :reek:FeatureEnvy
+def first_ex(a_side, b_side, c_side)
+  (a_side + b_side) <= c_side || (b_side + c_side) <= a_side || (c_side + a_side) <= b_side
+end
+
+# :reek:UtilityFunction and :reek:FeatureEnvy
+def triangle(a_side, b_side, c_side)
+  raise TriangleError if a_side <= 0 || b_side <= 0 || c_side <= 0
```
Comment (stahor) [inline 3444/0/triangle.rb]: there is `negative?`. You can also use `[].any?(&:negative?)`

File: 3444/0/triangle.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+#   :scalene      if no sides are equal
+#
+# The tests for this method can be found in
+#   about_triangle_project.rb
+# and
+#   about_triangle_project_2.rb
+#
+
+# :reek:UtilityFunction and :reek:FeatureEnvy
+def first_ex(a_side, b_side, c_side)
```
Comment (stahor) [inline 3444/0/triangle.rb]: first ex? WTF?

File: 3444/0/triangle.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  (a_side + b_side) <= c_side || (b_side + c_side) <= a_side || (c_side + a_side) <= b_side
```
Comment (stahor) [inline 3444/0/triangle.rb]: parentheses are not needed here actually. You can also make a method which will check the inequality for a single permutation

File: 3444/0/triangle.rb (hunk 7 above)
Comment (svintbel) [reply 3444/0/triangle.rb]: yea, it's looks horrible) thanks for review, need some sleep for going on

File: 3444/0/triangle.rb (hunk 8 above)
Comment (svintbel) [reply 3444/0/triangle.rb]: sorry, but i can't correct understand what u mean under ' method which will check the inequality for a single permutation'. I guess i need to add more abstraction?

Comment (stahor) [reply 3444/0/triangle.rb]: kinda
```
def satisfies_triangle_inequality?(side_a, side_b, side_c)
  side_a + side_b > side_c
end
```

File: 3444/0/.path_progress [hunk 9]
```diff
@@ -0,0 +1 @@
...108,108,108,108,108,110,111,113,113,113,114,115,115,115,115,115,115,115,115,115,116,120,120,120,121,121,121,122,124,124,124,124,126,127,128,132,132,133,136,137,139,141,141,142,142,144,145,148,152,154,155,155,155,156,156,158,158,158,159,159,160,160,160,160,160,160,161,167,168,168,169,170,175,175,175,175,179,180,180,182,182,182,182,181,182,182,182,182,182,182,182,182,184,184,184,184,184,184,184,185,186,185,186,186,186,186,186,185,185,185,185,185,185,185,186,185,184,186,186,186,186,186,186,185,185,184,186,186,187,187,187,187,187,185,185,185,185,185,185,185,185,186,186,185,185,185,186,186,187,187,187,187,187,187,185,184,187,187,192,192,193,193,194,194,194,194,194,194,194,194,194,194,194,194,194,194,194,195,195,195,197,202,204,206,207,207,207,207,207,207,208,211,212,212,212,212,212,212,212,212,212,212,212,212,212,212,212,212,212,212,212,212,216,214,216,212,216,216,216,219,222,222,223,228,228,234,236,236,236,236,237,237,246,255,256,258,260,261,261,261,261,262,262,264,265,267,267,214,267,268,268,268,268,270,270,270,270,270,271,271,271,271,271,271,277,271,271,277,278,280,281,281,212,182,187,187,184,184,186,267,178,50,50,49,50,50,50,50,212,185,184,185,185,185,189,189,189,189,189,152,152,160
```
Comment (stahor) [inline 3444/0/.path_progress]: neline

File: 3444/0/triangle.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+#
+# The tests for this method can be found in
+#   about_triangle_project.rb
+# and
+#   about_triangle_project_2.rb
+#
+
+# :reek:UtilityFunction and :reek:FeatureEnvy
+def satisfies_triangle_inequality?(a_side, b_side, c_side)
+  a_side + b_side <= c_side || b_side + c_side <= a_side || c_side + a_side <= b_side
```
Comment (stahor) [inline 3444/0/triangle.rb]: the idea was to extract a method for a single comparison. Not critical

Comment (svintbel) [reply 3444/0/triangle.rb]: oh, thanks, let it be so

File: 3444/0/triangle.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+end
+
+# :reek:UtilityFunction and :reek:FeatureEnvy
+def triangle(a_side, b_side, c_side)
+  raise TriangleError unless [a_side, b_side, c_side].all?(&:positive?)
+  raise TriangleError if satisfies_triangle_inequality? a_side, b_side, c_side
```
Comment (stahor) [inline 3444/0/triangle.rb]: Except for some rare cases you should never call methods without parentheses if you pass arguments there. How the hell did u pass rubocop check?

Comment (svintbel) [reply 3444/0/triangle.rb]: i don't know, i didn't disable this cop
Can fix this issue at few moments

Comment (svintbel) [reply 3444/0/triangle.rb]: i don't know, i didn't disable this cop
Can fix this issue at few moments

File: 3444/0/triangle.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+  if a_side == b_side && a_side == c_side
+    :equilateral
+  elsif a_side == b_side || a_side == c_side || b_side == c_side
```
Comment (stahor) [inline 3444/0/triangle.rb]: alternative solution `[a_side, b_side, c_side].uniq.size == 2` not really sure that it's better. Just wanted to show you the power of ruby.
You could use the number of uniq sides as argument for `case` statement btw. Or even transform that `case` into a hash

Comment (svintbel) [reply 3444/0/triangle.rb]: thanks, i guess i'll try to use this experience at next tasks

File: 3444/0/about_scoring_project.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+#
+# More scoring examples are given in the tests below:
+#
+# Your goal is to write the score method.
+# rubocop:disable Metrics/MethodLength
+# rubocop:disable Style/Next
+
+# :reek:UncommunicativeVariableName and :reek:TooManyStatements and :reek:UtilityFunction
+class Array
+  def delete_first(item)
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: o_O

Comment (svintbel) [reply 3444/0/about_scoring_project.rb]: i didn't find the same method in core

File: 3444/0/about_scoring_project.rb [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  if dice.length > 2
+    (1..6).each do |number|
+      if dice.count(number) > 2
+
+        total_score += triple_value[number]
+        3.times { dice.delete_first number }
+      end
+    end
+  end
+  dice.sum { |number| total_score += single_value[number] }
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: `total_score += dice.sum { |number| single_value[number] }`

File: 3444/0/about_scoring_project.rb [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+def score(dice)
+  total_score = 0
+  triple_value = [nil, 1000, 200, 300, 400, 500, 600]
+  single_value = [nil, 100, 0, 0, 0, 50, 0]
+  if dice.length > 2
+    (1..6).each do |number|
+      if dice.count(number) > 2
+
+        total_score += triple_value[number]
+        3.times { dice.delete_first number }
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: do not mutate arguments

File: 3444/0/about_scoring_project.rb [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+# :reek:UncommunicativeVariableName and :reek:TooManyStatements and :reek:UtilityFunction
+def score(dice)
+  total_score = 0
+  set = 0
+  triple_value = [nil, 1000, 200, 300, 400, 500, 600]
+  single_value = [nil, 100, 0, 0, 0, 50, 0]
+  if dice.length > 2
+    (1..6).each do |number|
+      if dice.count(number) > 2
+
```
Comment (stahor) [inline 3444/0/about_scoring_project.rb]: newline?

Comment (svintbel) [reply 3444/0/about_scoring_project.rb]: yea, i've added it only cause of rubocop or reek notes me about it

**Pull request merged**